File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        stack = [root]
        ancestor = TreeNode(float('inf'))
        if q.val < p.val:
            p,q = q,p
        while stack:
            node = stack.pop()
            if p.val <= node.val <= q.val and node.val <= ancestor.val and (node.right or node.left):
                ancestor = node
                return ancestor
                # if not stack:
                #     return ancestor
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)

        return ancestor

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int maxVal;
    private int count;

    public Solution(){
        this.maxVal = Integer.MIN_VALUE;
        this.count = 0;
    }
    private void preOrder(TreeNode node){
        if (node == null) return;
        int ancestorVal = node.val;
        if (node.val >= this.maxVal){
            this.maxVal = node.val;
            this.count++;
        }
        else ancestorVal = this.maxVal;

        preOrder(node.left);
        this.maxVal = ancestorVal;
        preOrder(node.right);
    }
    public int goodNodes(TreeNode root) {
        preOrder(root);
        return this.count;
    }
}
